aiida/manage/configuration/migrations/migrations.py

```diff
@@ -8,136 +8,282 @@
 
 
-class ConfigMigration:
-    """Defines a config migration."""
+class SingleMigration(Protocol):
+    """A single migration of the configuration."""
+    initial: int
+    """The initial configuration version."""
+    final: int
+    """The final configuration version."""
+
+    def upgrade(self, config: ConfigType) -> None:
+        """Migrate the configuration in-place."""
 
-    def __init__(self, migrate_function, version, version_oldest_compatible):
-        """Construct a ConfigMigration
+    def downgrade(self, config: ConfigType) -> None:
+        """Downgrade the configuration in-place."""
 
-        :param migrate_function: function which migrates the configuration dictionary
-        :param version: configuration version after the migration.
-        :param version_oldest_compatible: oldest compatible configuration version after the migration.
-        """
-        self.migrate_function = migrate_function
-        self.version = int(version)
-        self.version_oldest_compatible = int(version_oldest_compatible)
 
-    def apply(self, config):
-        """Apply the migration to the configuration."""
-        config = self.migrate_function(config)
-        config.setdefault('CONFIG_VERSION', {})['CURRENT'] = self.version
-        config.setdefault('CONFIG_VERSION', {})['OLDEST_COMPATIBLE'] = self.version_oldest_compatible
-        return config
+class Initial(SingleMigration):
+    """Add the required values for a new default profile.
 
+        * PROFILE_UUID
 
-def _1_add_profile_uuid(config):
+    The profile uuid will be used as a general purpose identifier for the profile, in
+    for example the RabbitMQ message queues and exchanges.
+    """
+    initial = 0
+    final = 1
+
+    def upgrade(self, config: ConfigType) -> None:
+        config.setdefault('CONFIG_VERSION', {})['OLDEST_COMPATIBLE'] = 0
+
+    def downgrade(self, config: ConfigType) -> None:
+        config.setdefault('CONFIG_VERSION', {})['OLDEST_COMPATIBLE'] = 0
+
+
+class AddProfileUuid(SingleMigration):
     """Add the required values for a new default profile.
 
         * PROFILE_UUID
 
     The profile uuid will be used as a general purpose identifier for the profile, in
     for example the RabbitMQ message queues and exchanges.
     """
-    for profile in config.get('profiles', {}).values():
-        from uuid import uuid4
-        profile['PROFILE_UUID'] = uuid4().hex
+    initial = 1
+    final = 2
 
-    return config
+    def upgrade(self, config: ConfigType) -> None:
+        from uuid import uuid4  # we require this import here, to patch it in the tests
+        for profile in config.get('profiles', {}).values():
+            profile.setdefault('PROFILE_UUID', uuid4().hex)
+        config.setdefault('CONFIG_VERSION', {})['OLDEST_COMPATIBLE'] = 0
 
+    def downgrade(self, config: ConfigType) -> None:
+        # leave the uuid present, so we could migrate back up
+        config.setdefault('CONFIG_VERSION', {})['OLDEST_COMPATIBLE'] = 0
```

Comment:
I see the point in not actually reverting the upgrade, because you would lose the UUID and if you upgrade again you get a different one. But isn't this a problem with the concept? Maybe in this case the older version will accept an additional key that its schema doesn't know about, but this may not be the case in general, would it?

Reply:
It is currently, and I think it should be enforced, that it always the case that schema allow for additional keys (this is the default in json schema), to allow for these "loseless" upgrades/downgrades.

In #5320, I have also removed the fact that `Profile` strips unknown keys, since I don't see any reason for it

Reply:
As commented below, fine to go with this behavior from now on and simply condone unknown/superfluous keys